src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -419,6 +518,13 @@ public NetworkPeer(IPEndPoint peerEndPoint, Network network, NetworkPeerConnecti
                 this.logger.LogTrace("(-)[EXCEPTION_CALLBACKS]");
                 throw;
             }
+            finally
+            {
+                if (this.onDisconnectedAsyncContext.Value.DisconnectCallbackRequested)
+                    this.onDisconnected(this);
```

Comment:
Effectively the only thing this does is delegate the dispose to the AsyncQueue, can we not write such a delegated dispose internally to the NetworkPeer?

Reply:
We can and it is exactly what it used to be before. 
And that leaded to the wrong implementation of IDisposable (by the end if Dispose() call the peer was not disposed)

In most cases Dispose() on network peer should be called from a separated task to avoid locks

Reply:
Hmm it will be a bit of an overhead as each peer will have to have its own dispose task waiting in the background so I see the reason of using a shared dispose queue. but this can be more specific to the NetworkPeer perhaps instead of injecting an OnDispose delegate just pass the DisposeHelper itself

Reply:
Right so if we just call `Task.Run(()=> this.DisposeInternal())` will that not work?

Reply:
It will work. But in order to actually dispose the peer we will need to: 
```
peer.Dispose()
peer.DisposeCompleted.Wait()
```

Which goes against the Dispose pattern

Reply:
What do you mean goes against the pattern?

Reply:
Dispose() suggests that by the end of it's call all threads that this object used are closed. 
If we do 
```
peer.Dispose()
peer.DisposeCompleted.Wait()
```

that basically means that we didn't finish disposing after exiting the Dispose() method and we would like to avoid that.